usr.bin/tip/tip.1
.\"	$NetBSD: tip.1,v 1.34.2.1 2025/08/02 05:58:44 perseant Exp $
.\"
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)tip.1	8.4 (Berkeley) 4/18/94
.\"
.Dd May 26, 2025
.Dt TIP 1
.Os
.Sh NAME
.Nm tip ,
.Nm cu
.Nd serial terminal emulator
.Sh SYNOPSIS
.Nm
.Op Fl v
.Fl Ns Ns Ar speed
.Ar system\-name
.Nm
.Op Fl v
.Fl Ns Ns Ar speed
.Ar phone\-number
.Nm cu
.Op Ar options
.Ar phone\-number
.Nm cu
.Op Ar options
.Dq Ar dir
.Nm cu
.Fl -help
.Sh DESCRIPTION
.Nm
and
.Nm cu
are used to connect to another system over a serial link.
In the era before modern networks, they were typically used to
connect to a modem in order to dial in to a remote host.
They are now frequently used for tasks such as attaching to the
serial console of another machine for administrative or
debugging purposes.
.Pp
The following option is available for
.Nm :
.Bl -tag -width 4n
.It Fl v
Set verbose mode.
.El
.Pp
The following options are available for
.Nm cu :
.Bl -tag -width 4n
.It Fl a Ar acu
Set the ACU port.
.It Fl c Ar number
Call this number.
.It Fl E Ar char
Use this escape character.
.It Fl e
Use even parity.
.It Fl F Ar flow
Set flow control to
.Ar hard ,
.Ar soft ,
or
.Ar none .
.It Fl f
Use no flow control.
.It Fl h
Echo characters locally (half-duplex mode).
.It Fl l Ar line
Specify the line to use.
Either of the forms like
.Pa tty00
or
.Pa /dev/tty00
are permitted.
.It Fl n
No escape (disable tilde).
.It Fl o
Use odd parity.
.It Fl P Ar parity
Set parity to
.Ar even ,
.Ar odd ,
or
.Ar none .
By default, no parity is used.
.It Fl p Ar acu
Set the ACU port.
.It Fl s Ar speed
Set the speed of the connection.
Defaults to 9600.
.It Fl t
Connect via a hard-wired connection to a host on a dial-up line.
.El
.Pp
If
.Ar speed
is specified it will override any baudrate specified in the system
description being used.
.Pp
If neither
.Ar speed
nor
.Ar system-name
are specified,
.Ar system-name
will be set to the value of the
.Ev HOST
environment variable.
.Pp
If
.Ar speed
is specified but
.Ar system-name
is not,
.Ar system-name
will be set to a value of
.Dq tip
with
.Ar speed
appended.
e.g.\&
.Ic tip -1200
will set
.Ar system-name
to
.Dq tip1200 .
.Pp
The optional
.Dq dir
argument is for Taylor implementation compatibility, indicating a
direct connection.
(This is how
.Nm cu
behaves by default, regardless.)
.Pp
Typed characters are normally transmitted directly to the remote
machine (which does the echoing as well).
A tilde (`~') appearing
as the first character of a line is an escape signal; the following
are recognized:
.Bl -tag -width flag
.It Ic \&~^D No or Ic \&~ .
Drop the connection and exit
(you may still be logged in on the
remote machine).
.It Ic \&~c Op Ar name
Change directory to
.Ar name
(no argument
implies change to your home directory).
.It Ic \&~!
Escape to a shell (exiting the shell will
return you to tip).
.It Ic \&~>
Copy file from local to remote.
.Nm
prompts for the name of a local file to transmit.
.It Ic \&~<
Copy file from remote to local.
.Nm
prompts first for the name of the file to be sent, then for
a command to be executed on the remote machine.
.It Ic \&~p Ar from Op Ar to
Send a file to a remote
.Ux
host.
The put command causes the remote
.Ux
system to run the command string ``cat > 'to''', while
.Nm
sends it the ``from''
file.
If the ``to'' file isn't specified the ``from'' file name is used.
This command is actually a
.Ux
specific version of the ``~>'' command.
.It Ic \&~t Ar from Op Ar to
Take a file from a remote
.Ux
host.
As in the put command the ``to'' file
defaults to the ``from'' file name if it isn't specified.
The remote host
executes the command string ``cat 'from';echo ^A'' to send the file to
.Nm .
.It Ic \&~|
Pipe the output from a remote command to a local
.Ux
process.
The command string sent to the local
.Ux
system is processed by the shell.
.It Ic \&~$
Pipe the output from a local
.Ux
process to the remote host.
The command string sent to the local
.Ux
system is processed by the shell.
.It Ic \&~C
Fork a child process on the local system to perform special protocols
such as
.Tn XMODEM .
The child program will be run with the following
arrangement of file descriptors:
.Bl -column -offset indent 0 "<->" "remote tty in"
.It 0 Ta <-> Ta remote tty in
.It 1 Ta <-> Ta remote tty out
.It 2 Ta <-> Ta local tty out
.El
.It Ic \&~+
Synonym for \&~C, provided for compatibility with other versions of
.Nm cu .
.It Ic \&~#
Send a
.Dv BREAK
to the remote system.
For systems which don't support the
necessary
.Ar ioctl
call the break is simulated by a sequence of line speed changes
and
.Dv DEL
characters.
.It Ic \&~s
Set a variable (see the discussion below).
.It Ic \&~^Z
Stop
.Nm
(only available with job control).
.It Ic \&~^Y
Stop only the ``local side'' of
.Nm
(only available with job control);
the ``remote side'' of
.Nm ,
the side that displays output from the remote host, is left running.
.It Ic \&~?
Get a summary of the tilde escapes
.El
.Pp
.Nm
uses the file
.Pa /etc/remote
to find how to reach a particular
system and to find out how it should operate while talking
to the system;
refer to
.Xr remote 5
for a full description.
Each system has a default baud rate with which to
establish a connection.
If this value is not suitable, the baud rate
to be used may be specified on the command line, e.g.
.Ql "tip -300 mds" .
.Pp
When
.Nm
establishes a connection it sends out a
connection message to the remote system; the default value, if any,
is defined in
.Pa /etc/remote
(see
.Xr remote 5 ) .
.Pp
When
.Nm
prompts for an argument (e.g. during setup of
a file transfer) the line typed may be edited with the standard
erase and kill characters.
A null line in response to a prompt,
or an interrupt, will abort the dialogue and return you to the
remote machine.
.Pp
.Nm
guards against multiple users connecting to a remote system
by opening modems and terminal lines with exclusive access,
and by honoring the locking protocol used by
.Xr uucico 8 .
.Pp
During file transfers
.Nm
provides a running count of the number of lines transferred.
When using the ~> and ~< commands, the ``eofread'' and ``eofwrite''
variables are used to recognize end-of-file when reading, and
specify end-of-file when writing (see below).
File transfers normally depend on tandem mode for flow control.
If the remote
system does not support tandem mode, ``echocheck'' may be set
to indicate
.Nm
should synchronize with the remote system on the echo of each
transmitted character.
.Pp
When
.Nm
must dial a phone number to connect to a system it will print
various messages indicating its actions.
.Nm
supports the
.Tn DEC DN Ns -11
and
Racal-Vadic 831 auto-call-units;
the
.Tn DEC DF Ns \&02
and
.Tn DF Ns \&03 ,
Ventel 212+, Racal-Vadic 3451, and
Bizcomp 1031 and 1032 integral call unit/modems.
.Ss VARIABLES
.Nm
maintains a set of
.Ar variables
which control its operation.
Some of these variables are read-only to normal users (root is allowed
to change anything of interest).
Variables may be displayed
and set through the ``s'' escape.
The syntax for variables is patterned
after
.Xr vi 1
and
.Xr mail 1 .
Supplying ``all''
as an argument to the set command displays all variables readable by
the user.
Alternatively, the user may request display of a particular
variable by attaching a `?' to the end.
For example ``escape?'' displays
the current escape character.
.Pp
Variables are numeric, string, character, or boolean values.
Boolean
variables are set merely by specifying their name; they may be reset
by prepending a `!' to the name.
Other variable types are set by
concatenating an `=' and the value.
The entire assignment must not
have any blanks in it.
A single set command may be used to interrogate
as well as set a number of variables.
Variables may be initialized at run time by placing set commands
(without the ``~s'' prefix in a file
.Pa .tiprc
in one's home directory).
The
.Fl v
option causes
.Nm
to display the sets as they are made.
Certain common variables have abbreviations.
The following is a list of common variables,
their abbreviations, and their default values.
.Bl -tag -width Ar
.It Ar beautify
(bool) Discard unprintable characters when a session is being scripted;
abbreviated
.Ar be  .
.It Ar baudrate
(num) The baud rate at which the connection was established;
abbreviated
.Ar ba  .
.It Ar dialtimeout
(num) When dialing a phone number, the time (in seconds)
to wait for a connection to be established; abbreviated
.Ar dial  .
.It Ar echocheck
(bool) Synchronize with the remote host during file transfer by
waiting for the echo of the last character transmitted; default is
.Ar off  .
.It Ar eofread
(str) The set of characters which signify an end-of-transmission
during a ~< file transfer command; abbreviated
.Ar eofr  .
.It Ar eofwrite
(str) The string sent to indicate end-of-transmission during
a ~> file transfer command; abbreviated
.Ar eofw  .
.It Ar eol
(str) The set of characters which indicate an end-of-line.
.Nm
will recognize escape characters only after an end-of-line.
.It Ar escape
(char) The command prefix (escape) character; abbreviated
.Ar es  ;
default value is `~'.
.It Ar exceptions
(str) The set of characters which should not be discarded
due to the beautification switch; abbreviated
.Ar ex  ;
default value is ``\et\en\ef\eb''.
.It Ar force
(char) The character used to force literal data transmission;
abbreviated
.Ar fo  ;
default value is `^P'.
.It Ar framesize
(num) The amount of data (in bytes) to buffer between file system
writes when receiving files; abbreviated
.Ar fr  .
.It Ar host
(str) The name of the host to which you are connected; abbreviated
.Ar ho  .
.It Ar prompt
(char) The character which indicates an end-of-line on the remote
host; abbreviated
.Ar pr  ;
default value is `\en'.
This value is used to synchronize during
data transfers.
The count of lines transferred during a file transfer
command is based on receipt of this character.
.It Ar raise
(bool) Upper case mapping mode; abbreviated
.Ar ra  ;
default value is
.Ar off  .
When this mode is enabled, all lower case letters will be mapped to
upper case by
.Nm
for transmission to the remote machine.
.It Ar raisechar
(char) The input character used to toggle upper case mapping mode;
abbreviated
.Ar rc  ;
default value is `^A'.
.It Ar record
(str) The name of the file in which a session script is recorded;
abbreviated
.Ar rec  ;
default value is ``tip.record''.
.It Ar script
(bool) Session scripting mode; abbreviated
.Ar sc  ;
default is
.Ar off  .
When
.Ar script
is
.Li true  ,
.Nm
will record everything transmitted by the remote machine in
the script record file specified in
.Ar record  .
If the
.Ar beautify
switch is on, only printable
.Tn ASCII
characters will be included in
the script file (those characters between 040 and 0177).
The
variable
.Ar exceptions
is used to indicate characters which are an exception to the normal
beautification rules.
.It Ar tabexpand
(bool) Expand tabs to spaces during file transfers; abbreviated
.Ar tab  ;
default value is
.Ar false .
Each tab is expanded to 8 spaces.
.It Ar tandem
(bool) Use XON/XOFF flow control to throttle data from the remote host;
abbreviated
.Ar ta .
The default value is
.Ar true
unless the
.Ar nt
capability has been specified in
.Pa /etc/remote ,
in which case the default value is
.Ar false .
.It Ar verbose
(bool) Verbose mode; abbreviated
.Ar verb  ;
default is
.Ar true  .
When verbose mode is enabled,
.Nm
prints messages while dialing, shows the current number
of lines transferred during a file transfer operations,
and more.
.El
.Sh ENVIRONMENT
.Nm
uses the following environment variables:
.Bl -tag -width Fl
.It Ev SHELL
(str) The name of the shell to use for the ~! command; default
value is ``/bin/sh'', or taken from the environment.
.It Ev HOME
(str) The home directory to use for the ~c command; default
value is taken from the environment.
.It Ev HOST
Check for a default host if none specified.
.El
.Pp
The variables
.Ev ${REMOTE}
and
.Ev ${PHONES}
are also exported.
.Sh FILES
.Bl -tag -width /var/spool/lock/LCK..* -compact
.It Pa /etc/remote
Global system descriptions.
.It Pa /etc/phones
Global phone number data base.
.It ${REMOTE}
Private system descriptions.
.It ${PHONES}
Private phone numbers.
.It ~/.tiprc
Initialization file.
.It Pa tip.record
Record file.
.El
.Sh DIAGNOSTICS
Diagnostics are, hopefully, self explanatory.
.Sh SEE ALSO
.Xr phones 5 ,
.Xr remote 5
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .
The
.Nm cu
.Pq Dq Call Unix
command first appeared outside of Bell Labs in PWB/UNIX 1.0.
.Bx
reimplemented it as part of the
.Nm
command.
.Sh BUGS
The full set of variables is undocumented and should, probably, be
pared down.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.3068.2.2 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 11.0 to NetBSD 12.0:




usr.bin/tip/tip.c
/*	$NetBSD: tip.c,v 1.63 2020/04/23 00:35:14 joerg Exp $	*/

/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#include <ctype.h>
#include <libgen.h>

#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1983, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)tip.c	8.1 (Berkeley) 6/6/93";
#endif
__RCSID("$NetBSD: tip.c,v 1.63 2020/04/23 00:35:14 joerg Exp $");
#endif /* not lint */

/*
 * tip - UNIX link to other systems
 *  tip [-v] [-speed] system-name
 * or
 *  cu [options] [phone-number|"dir"]
 */
#include "tip.h"
#include "pathnames.h"

struct termios	term;
struct termios	defterm;
struct termios	defchars;

FILE	*fscript;

int	attndes[2];
int	fildes[2];
int	repdes[2];
int	FD;
#ifndef __lint__  /* not used by hayes.c, but used by some other dialers */
int	AC;
#endif /*__lint__*/
int	sfd;
int	pid;
uid_t	uid, euid;
gid_t	gid, egid;
int	stop;
int	quit;
int	stoprompt;
int	timedout;
int	cumode;
int	bits8;
#define STRIP_PAR	(bits8 ? 0377 : 0177)

char	fname[80];
char	copyname[80];
char	ccc;

int	odisc;
int	vflag;

char	*DV;
char	*EL;
char	*CM;
char	*IE;
char	*OE;
char	*CU;
char	*AT;
char	*PN;
char	*DI;
char	*PA;

char	*PH;
char	*RM;
char	*HO;

long	BR;
long	FS;

long	DU;
long	HW;
char	*ES;
char	*EX;
char	*FO;
char	*RC;
char	*RE;
char	*PR;
long	DL;
long	CL;
long	ET;
long	HD;
char	DC;

__dead static void	tipusage(void);

int	escape(void);
__dead static void	intprompt(int);
__dead static void	tipin(void);

char	PNbuf[256];			/* This limits the size of a number */

static char path_phones[] = _PATH_PHONES;

int
main(int argc, char *argv[])
{
	char *System = NULL;
	int c, i;
	char *p;
	const char *q;
	char sbuf[12];
	int cmdlineBR;
	int fcarg;

	setprogname(argv[0]);
	gid = getgid();
	egid = getegid();
	uid = getuid();
	euid = geteuid();
	if (strcmp(getprogname(), "cu") == 0) {
		cumode = 1;
		cumain(argc, argv);
		goto cucommon;
	}

	if (argc > 4)
		tipusage();

	if (!isatty(0))
		errx(EXIT_FAILURE, "must be interactive");

	cmdlineBR = 0;
	while((c = getopt(argc, argv, "v0123456789")) != -1) {
		switch(c) {

		case 'v':
			vflag++;
			break;

		case '0': case '1': case '2': case '3': case '4':
		case '5': case '6': case '7': case '8': case '9':
			cmdlineBR = cmdlineBR * 10 + (c - '0');
			BR = cmdlineBR;
			break;

		default:
			warnx("%s, unknown option", argv[1]);
			break;
		}
	}

	argc -= optind;
	argv += optind;

	if (argc != 1)
		tipusage();
	else
		System = argv[0];


	if (System == NULL)
		goto notnumber;
	if (isalpha((unsigned char)*System))
		goto notnumber;
	/*
	 * System name is really a phone number...
	 * Copy the number then stomp on the original (in case the number
	 *	is private, we don't want 'ps' or 'w' to find it).
	 */
	if (strlen(System) > sizeof PNbuf - 1) {
		errx(1, "phone number too long (max = %d bytes)",
			(int)sizeof(PNbuf) - 1);
	}
	(void)strlcpy(PNbuf, System, sizeof(PNbuf));
	for (p = System; *p; p++)
		*p = '\0';
	PN = PNbuf;
	(void)snprintf(sbuf, sizeof sbuf, "tip%d", (int)BR);
	System = sbuf;

notnumber:
	(void)signal(SIGINT, cleanup);
	(void)signal(SIGQUIT, cleanup);
	(void)signal(SIGHUP, cleanup);
	(void)signal(SIGTERM, cleanup);

	if ((i = hunt(System)) == 0) {
		errx(3, "all ports busy");
	}
	if (i == -1) {
		errx(3, "link down");
	}
	setbuf(stdout, NULL);

	/*
	 * Kludge, their's no easy way to get the initialization
	 *   in the right order, so force it here
	 */
	if ((PH = getenv("PHONES")) == NULL)
		PH = path_phones;
	vinit();				/* init variables */
	setparity("none");			/* set the parity table */

	/*
	 * Hardwired connections require the
	 *  line speed set before they make any transmissions
	 *  (this is particularly true of things like a DF03-AC)
	 */
	if (HW) {
		if (ttysetup((speed_t)number(value(BAUDRATE))) != 0) {
			errx(3, "bad baud rate %d",
			    (int)number(value(BAUDRATE)));
		}
	}
	if ((q = tip_connect()) != NULL) {
		errx(1, "\07%s\n[EOT]", q);
	}
	if (!HW) {
		if (ttysetup((speed_t)number(value(BAUDRATE))) != 0) {
			errx(3, "bad baud rate %d",
			    (int)number(value(BAUDRATE)));
		}
	}


cucommon:
	/*
	 * From here down the code is shared with
	 * the "cu" version of tip.
	 */

	/*
	 * Direct connections with no carrier require using O_NONBLOCK on
	 * open, but we don't want to keep O_NONBLOCK after open because it
	 * will cause busy waits.
	 */
	if (DC &&
	    ((fcarg = fcntl(FD, F_GETFL, 0)) < 0 ||
	     fcntl(FD, F_SETFL, fcarg & ~O_NONBLOCK) < 0)) {
		err(1, "can't clear O_NONBLOCK");
	}

	(void)tcgetattr(0, &defterm);
	term = defterm;
	term.c_lflag &= ~(ICANON|IEXTEN|ECHO);
	term.c_iflag &= ~(INPCK|ICRNL);
	term.c_oflag &= ~OPOST;
	term.c_cc[VMIN] = 1;
	term.c_cc[VTIME] = 0;
	defchars = term;
	term.c_cc[VINTR] = term.c_cc[VQUIT] = term.c_cc[VSUSP] =
		term.c_cc[VDSUSP] = term.c_cc[VDISCARD] =
	 	term.c_cc[VLNEXT] = _POSIX_VDISABLE;
	raw();

	(void)pipe(attndes);
	(void)pipe(fildes);
	(void)pipe(repdes);
	(void)signal(SIGALRM, alrmtimeout);

	/*
	 * Everything's set up now:
	 *	connection established (hardwired or dialup)
	 *	line conditioned (baud rate, mode, etc.)
	 *	internal data structures (variables)
	 * so, fork one process for local side and one for remote.
	 */
	(void)printf("%s", cumode ? "Connected\r\n" : "\07connected\r\n");
	switch (pid = fork()) {
	default:
		tipin();
		break;
	case 0:
		tipout();
		break;
	case -1:
		err(1, "can't fork");
	}
	/*NOTREACHED*/
	exit(0);	/* XXX: pacify gcc */
}

void
tipusage(void)
{
	(void)fprintf(stderr, "usage: %s [-v] [-speed] system-name\n",
	    getprogname());
	(void)fprintf(stderr, "       %s [-v] [-speed] phone-number\n",
	    getprogname());
	exit(1);
}

void
/*ARGSUSED*/
cleanup(int dummy __unused)
{

	if (odisc)
		(void)ioctl(0, TIOCSETD, &odisc);
	_exit(0);
}

/*
 * put the controlling keyboard into raw mode
 */
void
raw(void)
{

	(void)tcsetattr(0, TCSADRAIN, &term);
}


/*
 * return keyboard to normal mode
 */
void
unraw(void)
{

	(void)tcsetattr(0, TCSADRAIN, &defterm);
}

static	jmp_buf promptbuf;

/*
 * Print string ``s'', then read a string
 *  in from the terminal.  Handles signals & allows use of
 *  normal erase and kill characters.
 */
int
prompt(const char *s, char *volatile p, size_t l)
{
	int c;
	char *b = p;
	sig_t oint, oquit;

	stoprompt = 0;
	oint = signal(SIGINT, intprompt);
	oquit = signal(SIGQUIT, SIG_IGN);
	unraw();
	(void)printf("%s", s);
	if (setjmp(promptbuf) == 0)
		while ((c = getchar()) != EOF && (*p = c) != '\n' &&
		    b + l > p)
			p++;
	*p = '\0';

	raw();
	(void)signal(SIGINT, oint);
	(void)signal(SIGQUIT, oquit);
	return (stoprompt || p == b);
}

/*
 * Interrupt service routine during prompting
 */
static void
/*ARGSUSED*/
intprompt(int dummy __unused)
{

	(void)signal(SIGINT, SIG_IGN);
	stoprompt = 1;
	(void)printf("\r\n");
	longjmp(promptbuf, 1);
}

/*
 * getchar() wrapper that checks for EOF on the local end.
 */
static int
xgetchar(void)
{
	int c = getchar();
	if (__predict_false(c == EOF)) {
		cleanup(SIGHUP);
		/* NOTREACHED */
	}

	return c & STRIP_PAR;
}


/*
 * ****TIPIN   TIPIN****
 */
static void
tipin(void)
{
	char gch, bol = 1;

	/*
	 * Kinda klugey here...
	 *   check for scripting being turned on from the .tiprc file,
	 *   but be careful about just using setscript(), as we may
	 *   send a SIGEMT before tipout has a chance to set up catching
	 *   it; so wait a second, then setscript()
	 */
	if (boolean(value(SCRIPT))) {
		(void)sleep(1);
		setscript();
	}

	for (;;) {
		gch = xgetchar();
		if ((gch == character(value(ESCAPE))) && bol) {
			if (!(gch = escape()))
				continue;
		} else if (!cumode &&
		    gch && gch == character(value(RAISECHAR))) {
			setboolean(value(RAISE), !boolean(value(RAISE)));
			continue;
		} else if (gch == '\r' || gch == '\n') {
			bol = 1;
			xpwrite(FD, &gch, 1);
			if (boolean(value(HALFDUPLEX)))
				(void)printf("%s\n", gch == '\r' ? "\r" : "");
			continue;
		} else if (!cumode && gch && gch == character(value(FORCE)))
			gch = xgetchar();
		bol = any(gch, value(EOL));
		if (boolean(value(RAISE)) && islower((unsigned char)gch))
			gch = toupper((unsigned char)gch);
		xpwrite(FD, &gch, 1);
		if (boolean(value(HALFDUPLEX)))
			(void)printf("%c", gch);
	}
}

/*
 * Escape handler --
 *  called on recognition of ``escapec'' at the beginning of a line
 */
int
escape(void)
{
	char gch;
	esctable_t *p;
	char c = character(value(ESCAPE));

	gch = xgetchar();
	for (p = etable; p->e_char; p++)
		if (p->e_char == gch) {
			if ((p->e_flags&PRIV) && uid)
				continue;
			(void)printf("%s", ctrl(c));
			(*p->e_func)(gch);
			return (0);
		}
	/* ESCAPE ESCAPE forces ESCAPE */
	if (c != gch)
		xpwrite(FD, &c, 1);
	return (gch);
}

int
any(char c, const char *p)
{

	while (p && *p)
		if (*p++ == c)
			return (1);
	return (0);
}

char *
interp(const char *s)
{
	static char buf[256];
	char *p = buf, c;
	const char *q;

	while ((c = *s++) != 0 && buf + sizeof buf - p > 2) {
		for (q = "\nn\rr\tt\ff\033E\bb"; *q; q++)
			if (*q++ == c) {
				*p++ = '\\'; *p++ = *q;
				goto next;
			}
		if (c < 040) {
			*p++ = '^'; *p++ = c + 'A'-1;
		} else if (c == 0177) {
			*p++ = '^'; *p++ = '?';
		} else
			*p++ = c;
	next:
		;
	}
	*p = '\0';
	return (buf);
}

char *
ctrl(char c)
{
	static char s[3];

	if (c < 040 || c == 0177) {
		s[0] = '^';
		s[1] = c == 0177 ? '?' : c+'A'-1;
		s[2] = '\0';
	} else {
		s[0] = c;
		s[1] = '\0';
	}
	return (s);
}

/*
 * Help command
 */
void
help(char c)
{
	esctable_t *p;

	(void)printf("%c\r\n", c);
	for (p = etable; p->e_char; p++) {
		if ((p->e_flags&PRIV) && uid)
			continue;
		(void)printf("%2s", ctrl(character(value(ESCAPE))));
		(void)printf("%-2s %c   %s\r\n", ctrl(p->e_char),
			p->e_flags&EXP ? '*': ' ', p->e_help);
	}
}

/*
 * Set up the "remote" tty's state
 */
int
ttysetup(speed_t spd)
{
	struct termios	cntrl;

	(void)tcgetattr(FD, &cntrl);
	(void)cfsetospeed(&cntrl, spd);
	(void)cfsetispeed(&cntrl, spd);
	cntrl.c_cflag &= ~(CSIZE|PARENB);
	cntrl.c_cflag |= CS8;
	if (DC)
		cntrl.c_cflag |= CLOCAL;
	if (boolean(value(HARDWAREFLOW)))
		cntrl.c_cflag |= CRTSCTS;
	cntrl.c_iflag &= ~(ISTRIP|ICRNL);
	cntrl.c_oflag &= ~OPOST;
	cntrl.c_lflag &= ~(ICANON|ISIG|IEXTEN|ECHO);
	cntrl.c_cc[VMIN] = 1;
	cntrl.c_cc[VTIME] = 0;
	if (boolean(value(TAND)))
		cntrl.c_iflag |= IXOFF|IXON;
	else
		cntrl.c_iflag &= ~(IXOFF|IXON);
	return tcsetattr(FD, TCSAFLUSH, &cntrl);
}

static char partab[0200];

/*
 * Do a write to the remote machine with the correct parity.
 * We are doing 8 bit wide output, so we just generate a character
 * with the right parity and output it.
 */
void
xpwrite(int fd, char *buf, size_t n)
{
	size_t i;
	char *bp;

	bp = buf;
	if (bits8 == 0)
		for (i = 0; i < n; i++) {
			*bp = partab[(*bp) & 0177];
			bp++;
		}
	if (write(fd, buf, n) < 0) {
		if (errno == EIO)
			tipabort("Lost carrier.");
		/* this is questionable */
		warn("write");
	}
}

/*
 * Build a parity table with appropriate high-order bit.
 */
void
setparity(const char *defparity)
{
	int i, flip, clr, set;
	const char *parity;
	static char *curpar;

	if (value(PARITY) == NULL || ((char *)value(PARITY))[0] == '\0') {
		if (curpar != NULL)
			free(curpar);
		value(PARITY) = curpar = strdup(defparity);
	}
	parity = value(PARITY);
	if (strcmp(parity, "none") == 0) {
		bits8 = 1;
		return;
	}
	bits8 = 0;
	flip = 0;
	clr = 0377;
	set = 0;
	if (strcmp(parity, "odd") == 0)
		flip = 0200;			/* reverse bit 7 */
	else if (strcmp(parity, "zero") == 0)
		clr = 0177;			/* turn off bit 7 */
	else if (strcmp(parity, "one") == 0)
		set = 0200;			/* turn on bit 7 */
	else if (strcmp(parity, "even") != 0) {
		(void)fprintf(stderr, "%s: unknown parity value\r\n", parity);
		(void)fflush(stderr);
	}
	for (i = 0; i < 0200; i++)
		partab[i] = ((evenpartab[i] ^ flip) | set) & clr;
}
